### ISSUE TITLE
Add performance ranking scores / grades to profile page

### ISSUE THREAD

**Issue by brendanheywood** [theCrag/website#2505]
So we can add a new 'performance' grade which would probably replace the 'onsight' 

![image](https://cloud.githubusercontent.com/assets/187449/21641746/eeabdc1a-d2d1-11e6-8e6f-2e5f568b4049.png)

The challenge with this is we want it as a grade, which means we need to choose a grade system. We need to do this consistently based on the profile we are looking at, so grab their country, then map that to a context, then map it to a system. For every ranking add this extra info, so instead of:

```
"rating": {
"sport": 2914,
"trad": 2535,
"boulder": 3370
},
```

we'd have something vaguely like:

```
"rating": {
  "decaysportpower" : {
    "score": 2914,
    "context": "AU",
    "system": "Ewbanks",
    "grade": "24"
  },
  "decaytradpower": {
    "score": 2535,
    "context": "AU",
    "system": "Ewbanks",
    "grade": "21"
  },
  "decayboulderpower": {
    "score": 3370,
    "context": "AU",
    "system": "V-scale",
    "grade": "V5"
  }
},
```

This will work fine for the profile pages, but we also want to use this stuff on the ranking pages and it's really ugly and weird to see each climber using a different grade system. So we also need to infer a grade content and system for a collection of climbers. For any ranking at a node this is easy as we can just grab the context of that node, but at the world level it's not so trivial. I think in this case we should default back to the grade context of country of the user viewing the ranking. If the user is anonymous then we I think for now we just default to either US or FR. Ultimately this needs to be something we can set in the session / account prefs.

**Comment by scd** [comment]
Are we branding the rating number anywhere?

**Comment by brendanheywood** [comment]
As I've said elsewhere I'm not a fan of even showing the raw score, but concede we should show this on the ranking pages with whatever branding we want. I don't think this is meaningful in isolation on a profile page so I wouldn't show it there.

**Comment by scd** [comment]
Note that there is a name clash with the existing performance grade measure. I am sure we will end up just turning the old performance grade off and replace, but I doubt we will be ready for that for a beta trial release.

**Comment by brendanheywood** [comment]
We will want them side by side for a while so just call it something slightly different

**Comment by scd** [comment]
Note to self when I jump back onto this issue. See the 

```
account-power-rating-06-climber-grade-system-preference.sql
```
As a demonstration script for getting grading system preferences.

**Comment by scd** [comment]
This is what I currently have in the profile page template data:

```
"rating": {

    "trad": {
        "total": 3484,
        "totalGrade": "28",
        "decay": 3483,
        "system": "Ewbanks",
        "decayGrade": "28"
    },
    "sport": {
        "total": 3164,
        "totalGrade": "26",
        "decay": 1593,
        "system": "Ewbanks",
        "decayGrade": "16"
    },
    "boulder": {
        "total": 2420,
        "totalGrade": "V2",
        "decay": 2207,
        "system": "V-Scale",
        "decayGrade": "V1"
    }

},
```

Do we need context? I don't think context helps at all here.

Also do we need grade band?

**Comment by brendanheywood** [comment]
Yes I absolutely want the context but maybe here isn't actually the place because the context should be the same for all 3 scores? There is a lot of overlap here with #2526 where I want to clearly show what grade systems a users profile is in. I really want the context 

But there is a potential complication, a person might have lived in Britain and climbed heaps of trad, and then moved to France and done heaps of bouldering. So if we choose the most ascents in each category you might end up with a UK > UK Trad context / system for one and a FR > Font for the other. This is probably an edge case, but I'd actually prefer that for an entire persons profile we dice which context it should be based on all of their ascents, and then everything on the entire profile including the rating converted to a grade, and all the other grades is shown relative to that context.

Note that I'm not talking about converting the grade over, just showing them relative, so in the case above lets say that the trad rating ended up as UK:6a (I think we are discarding the adjectval rating?) and the bouldering rating ended up as FB:6a. Right now on lots of profiles we have issues like this where you don't know what system you are looking at. So lets pick whatever algorithm to pick the profile context, say just biggest proportion of ticks, and let says the bouldering in FR won. So in the top left corner of the account page we have the context exactly the same as we do on the index pages, the french bouldering grades anywhere on the page would just appear as '6a', any french routes would show as 'FR:6a' and the UK trad routes as 'UK:6a'

The account page template has a 'defaultSystem' but I'm sure how it's calculated, and there is doesn't appear to be a context. In any case it's barely used anywhere. Ultimately I'd like this to be largely automatic but have an account setting so that someone can explicitly set their profile context (see #2540). And then on top of that the person viewing their page can also override it and also convert everything (see #2526)

**Comment by brendanheywood** [comment]
another thing that we need to decide on is which of the 3 ratings do we show front and center? Just using the most ticks seems simplistic, and I'm not sure what the best balance is here.

We currently show the hardest grade of all time, and then onsight. Perhaps we leave hardest because it's the most flattering, and then next to it put the performance grade? And try to match up the performance grade that best matches the hardest? ie I'd probably be hardest V7 and current performing at V5.

I'm starting to second guess what we should do here

**Comment by scd** [comment]
Grading systems can be in multiple contexts and there are multiple grading systems for each context. The main purpose is for parsing grades so there are no conflicts. You might be ahead of me with your thought patterns here, so I will need some time to get up to speed on the contexts.

If I understand it correctly you want to make sure we pick related grading systems not just the the most common grading system. 

The best way to get the default context for a user is to use the country. Each country has a default context. We have also required that particular nodes can override the default context and change it for that area.

 > another thing that we need to decide on is which of the 3 ratings do we show front and center

I am not too worried about this because this is something we can refine the business rules for as we go. I am pretty sure we will not get this right first go.

 > Perhaps we leave hardest because it's the most flattering

Yes, the profile headline grade should be your hardest. We are allowed to use any statistic we want here, and we should use the most flattering. We do not have to use the new stats at the profile headline. 

> I'd probably be hardest V7 and current performing at V5

That seems reasonable. I think we should be flexible here. If somebody has not climbed for 10 years then there performance grade may have gone down from a V7 to a V2, in which case we should not show V2 in the profile page but maybe something <V5 or just a symbol like '?'.

Note that I think the 'Onsights' statistic is actually a performance grade, so we could use a version of our new stat for that.

**Comment by scd** [comment]
I have added grade context to the rating stuff for the profile page. So I now get:

```
rating": {

    "sport": {
        "totalGrade": "27",
        "system": "Ewbanks",
        "total": 3334,
        "decayGrade": "15",
        "decay": 1493
    },
    "boulder": {
        "totalGrade": "V2",
        "system": "V-Scale",
        "total": 2420,
        "decayGrade": "V1",
        "decay": 2204
    },
    "trad": {
        "totalGrade": "36",
        "system": "Ewbanks",
        "total": 4466,
        "decayGrade": "36",
        "decay": 4465
    },
    "context": {
        "systems": [
            "AU",
            "BLDV",
            "AUAID",
            "STAR",
            "PROT"
        ],
        "contextID": "7516384",
        "contextName": "Australian (Ewbanks ...)",
        "context": "AU"
    }

},
```

The algorithm is as such:
  * Based on the sum of all the ascents in the gear style search get the context with the maximum number of ascents
  * If equal (eg 10 ascents in US and 10 ascents in AU contexts) this ends up being the one with the smallest context id - not ideal, but ok for current purposes
 * If no ascents use country default
 * If still nothing use global default (US)

If you do an equal number of ascents in two different grading systems then I think I need to do more work to find the right one. This case may occur quite a lot, for example if you do 10 V-Scale boulder problems only then this applies to multiple contexts, and will result in them all having equal counts.

Anyway this issue is just an improvement to the data quality which I will do sometime after we confirm the template data is what is needed for the profile page.
